### PR TITLE
K8smetrics command

### PR DESCRIPTION
**Pull request by dmitry-onishchenko**
- add few methods to get k8s metrics
- add k8smetrics command to the client

### CODE REVIEW

**Comment by bgokden** [comment]
I am not sure if you can get the labels in metrics, we need to try if it is available.

File: kubernetes/metrics.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+				CPUf    float64 // field for storing processed CPU data
+				Memory  string  `json:"memory"`
+				MemoryF float64 // field for storing processed Memory data
+			} `json:"usage"`
+		} `json:"containers"`
+	} `json:"items"`
+}
+
+// Metrics returns list of metrics for a given namespace
+func Metrics(configPath string, namespace string, pods *PodMetricsList) error {
```
[the remark this answers is not in the document]
Comment (dmitry-onishchenko) [reply kubernetes/metrics.go]: Renamed

File: cmd/metrics.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/magneticio/vampkubistcli/kubernetes"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var namespace string
+
+// bootstrapCmd represents the bootstrap command
+var metricsCmd = &cobra.Command{
+	Use:   "metrics",
```
[the remark this answers is not in the document]
Comment (dmitry-onishchenko) [reply cmd/metrics.go]: Renamed to k8s-metrics

File: kubernetes/metrics.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	APIVersion string `json:"apiVersion"`
+	Metadata   struct {
+		SelfLink string `json:"selfLink"`
+	} `json:"metadata"`
+	Items []struct {
+		Metadata struct {
+			Name              string    `json:"name"`
+			Namespace         string    `json:"namespace"`
+			SelfLink          string    `json:"selfLink"`
+			CreationTimestamp time.Time `json:"creationTimestamp"`
```
[the remark this answers is not in the document]
Comment (dmitry-onishchenko) [reply kubernetes/metrics.go]: As for now I'm checking response from metrics server and if there are no labels, the method make additional query to k8s to get labels for each pod.

File: kubernetes/metrics.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+			}
+		}
+		res[i].CPU = sumCPU / float64(len(pods.Items[i].Containers))
+		res[i].Memory = sumMem / float64(len(pods.Items[i].Containers))
+	}
+	return res, nil
+}
+
+// ProcessMetrics converts string values for CPU and memory to float ones and stores them into dedicated fields
+func ProcessMetrics(stract interface{}) {
```
Comment (bgokden) [inline kubernetes/metrics.go]: stract or struct?

File: kubernetes/metrics.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+		res[i].Name = pods.Items[i].Metadata.Name
+		res[i].Labels = pods.Items[i].Metadata.Labels
+		var sumCPU, sumMem float64
+		for j := 0; j < len(pods.Items[i].Containers); j++ {
+			cpu, err := ConvertCPU(pods.Items[i].Containers[j].Usage.CPU)
+			if err == nil {
+				sumCPU += cpu
+			} else {
+				logging.Error("Conversion of CPU for %v failed - %v", pods.Items[i], err)
+				return nil, err
```
Comment (bgokden) [inline kubernetes/metrics.go]: I think returning error on one conversion can be too much, on startups of pods and similar events, there is usually empty or 0 values.

File: kubernetes/metrics.go (hunk 4 above)
Comment (dmitry-onishchenko) [reply kubernetes/metrics.go]: it's name of parameter and it cannot be reserved keyword (e.g. struct). Basically I wanted to show that it's intended to be struct internally that is passed as interface. But might be it's confusing.

Comment (dmitry-onishchenko) [reply kubernetes/metrics.go]: I'll rename it

**Comment by dmitry-onishchenko** [comment]
> This is good for me the way that it is now but we need to simplify this to use inside the adapter.
> We can simplify after integrating.
> We will probably remove unused fields to decrease memory usage.

@bgokden , Basically I implemented 2 ways of metrics processing. The first one (it's implemented by option `--raw` is just keeping all the fields from metric server response as is plus adding float representations for cpu and memory plus adding labels. The second way (it's implemented by option `--average` returns new structure that is filled up with average cpu and memory, labels and names of pods. If we are sure what we really need we can just remove obsolete way or use only one preferred way of getting metrics from the two in the adapter.

File: kubernetes/metrics.go (hunk 5 above)
Comment (dmitry-onishchenko) [reply kubernetes/metrics.go]: Fixed

File: kubernetes/metrics.go (hunk 4 above)
Comment (dmitry-onishchenko) [reply kubernetes/metrics.go]: Renamed

**Comment by dmitry-onishchenko** [comment]
@bgokden , please have a look - I've addressed raised concerns.

**Pull request merged**